### DISCUSSION TITLE
dev branch?

### DISCUSSION THREAD

**Discussion by ljleb** [Mikubill/sd-webui-controlnet#1699] [General]
Instead of merging PRs directly to main, it could make a smoother transition to merge to an intermediate dev branch so that more people have the chance to test new releases before they are published to the main branch.

This was discussed previously here: https://github.com/Mikubill/sd-webui-controlnet/discussions/604

**Comment by huchenlei** [comment]
The problem is how do we incentivise a stable user population to use the dev version of the extension and catch bugs. If we don't have a big enough population actively using the extension, the bug may still not be caught before releasing into main.

I think instead of doing a dev branch, we should offer a way to easily go back to previous "stable version". So that if something breaks in the newest commit, the user always have a fallback. In Chrome, there is a mechanism that if consistent crash is monitored after the push of an update, it will try to fallback to the version before the update.

**Reply by ljleb** [reply]
There used to be at least 2-3 people willing to help on this front, however this was a while ago (a couple months at least), so it may be too late to take the help of these people. I would not be surprized if we found other people willing to lend a hand, as it is in the interest of some users at least to test releases before they are served.

In the case of the webui, I am not sure if rolling back is going to work 100% of the time or create suprising situations. The code already struggles to update forward without any issue, I would think going backwards will be another can of worms.

Despite this, I agree with you that such a mechanims could be handy if it is implemented. In any case, there needs to be a discussion here about getting the unpredictable nature of updates under control.

**Reply by huchenlei** [reply]
I would suggest banning dynamic reloading of modules, at least not doing it partially like what we are doing now. If we cannot consistently refreshing every controlnet module loaded, there is a chance of inconsistent versions of modules running together.

We have seen plenty of this issue:
- Newly added function/class not found
- Type checking breaking because of the instance of old class persisted somewhere (This issue actually cannot be solved even we reload every module, as there can be stale instances stored somewhere else. For example: #1693 describes an issue where deforum extension conflicts with module reloading.)
- etc

About more stability on new commits, I believe expanding test coverage is the key. Currently our testcase heavily relies on unittest and api test, which cannot detect UI side changes. I think we should add selenium-based WebUI unittest that runs on upload to run diverse test scenarios.

**Reply by ljleb** [reply]
I think the partial reloading is the root cause of many unexpected issues. At the moment, when updating existing modules, adding importlib.reload for these modules is important.

Another cause for the failing typechecking is that other extensions also instantiate the external_code module, creating a completely unrelated set of classes and functions to the one used directly by controlnet. (for example, the ControlNetUnit class will have multiple versions during runtime if other extensions import controlnet)

There should be a way to list every python module of the controlnet extension. If it is possible to list them all automatically, then it should be just a matter of looping over the modules and reloading them all. Achieving this could fix a number of issues when updating without fully restarting the webui.

**Reply by ljleb** [reply]
Maybe it is possible to use this instead of importlib.reload? https://docs.python.org/3/library/sys.html#sys.modules

**Reply by ljleb** [reply]
Just to be clear, I don't know if this is a good idea. I'm just suggesting ideas here. I think if a reload of all modules can be done, then the webui host should be responsible for reloading extensions properly.

**Reply by lllyasviel** [reply]
https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/59419bd64a1581caccaac04dceb66c1c069a2db1/CHANGELOG.md?plain=1#L28

**Reply by ljleb** [reply]
This is for 1.4.0 though, which has not yet been merged in main. do you know if it only applies to installing extensions or if it also applies to extension updating?

**Reply by catboxanon** [reply]
Installing or updating both require a restart so it applies to both.
